### PR TITLE
Enhancement/chat UI md support

### PR DESCRIPTION
**Pull request by osala-eng**
## Description

Please include a summary of the change, e.g. what the new feature # is and/or what bug # it fixes. Please also include relevant motivation and context. List any dependencies that are required for this change.

Fixes/Implements #(issue/feature)

## Type of change

- [ ] Bug fix (non-breaking change which fixes an issue)
- [x] New feature (non-breaking change which adds functionality)
- [ ] Refactor and code improvement (non-breaking change which improves code quality and/or performance)
- [ ] Breaking change (fix or feature that would cause existing functionality to not work as expected)
- [ ] Documentation
- [ ] Tests
- [ ] Other chores such as maintenance

## How Has This Been Tested?

Please describe the tests that you ran to verify your changes. Provide instructions so we can reproduce. Please also list any relevant details for your test configuration by modifying the list below.

- [ ] Test A
- [ ] Test B
- [ ] Test C

**Test Configuration**:

Please describe the test setup. List them below as bullet points.

## Checklist:

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have commented my code in hard-to-understand areas
- [ ] I have made corresponding changes to the documentation
- [ ] My changes generate no new warnings
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes
- [ ] Any dependent changes have been merged and published in downstream modules
- [ ] I have checked my code and corrected any misspellings
- [ ] The commit message follows the convention of this project

### CODE REVIEW

**Comment by osala-eng** [comment]
I've modified the source code of `streamlit-chat` to support markdown formatting and published the new module as `streamlit-chat-plus` source code https://github.com/osala-eng/st-chat-plus

### Sample output
chat-new

**Comment by cwang** [comment]
Nice one. How about this - let's open source it here in `docqai` github org. Could you go ahead and fork the original repo st-chat and apply your changes here pls, just like what you did with st-chat-plus in your own github org. 

We can then shout about it for others to use it, as a way to market other stuff we build.

**Comment by cwang** [comment]
Actually looks like st-chat already supports html, could you double check before forking?

**Comment by cwang** [comment]
I created the fork anyway but looks like they added HTML support a few days ago :)

**Comment by osala-eng** [comment]
> I created the fork anyway but looks like they added HTML support a few days ago :) Let me know what you want to do next - perhaps keep it simple for now, i.e. avoiding forking?

I see, `st-chat` added support for markdown and I agree it's better to keep it simple and and avoid forking for now.

**Comment by cwang** [comment]
Maybe use this PR to see if you could make the sources linkable - right now they're listed as plain text.

**Comment by osala-eng** [comment]
> Maybe use this PR to see if you could make the sources linkable - right now they're listed as plain text.

Ok, let me get on this.

**Comment by osala-eng** [comment]
> Maybe use this PR to see if you could make the sources linkable - right now they're listed as plain text.

@cwang @janaka 


In order to make the sources linkable and enable users to preview documents in their browser using public doc viewers from Google or Microsoft, we should consider implementing a more secure approach for serving static files.

- Currently, we can preview PDF files easily since most web browsers can natively open them directly if sent to the frontend as an `arrayBuffer`.

- Inorder to view files such as `docx` and `xslx` we need to utilize a specialized document viewer API: There are third-party APIs available, such as Google Docs API or Microsoft Office Viewer API, which provide document rendering capabilities. By integrating with these APIs, we can securely pass the document file to the API and receive an embeddable HTML code or an iframe to display the document within our application. This approach allows for fine-grained control over access and maintains the security of the files.

#### Possible solutions
- One way in which we can serve the files using streamlit `static file serving` https://docs.streamlit.io/library/advanced-features/static-file-serving, to serve user documents from `.persited/upload` directory which posses a security risk.
- An alternative more secure approach could be: Setting up a reverse proxy, such as Nginx or HAProxy, can act as an intermediary between the `frontend` and the `backend` application server. This way, the reverse proxy can handle requests for static files directly, allowing us to configure security measures like access controls, rate limiting, and SSL/TLS termination.
- Leveraging a cloud storage service: Storing static files in a cloud storage solution like Amazon S3, Google Cloud Storage, or Azure Blob Storage can provide a scalable and secure way to serve these assets. We can then utilize their respective capabilities to generate signed URLs or configure access controls to ensure only authorized users can access the files.


What is your opinion on this and way foward regarding this?

**Comment by cwang** [comment]
Great points. Let me clarify:
- No need to *view* documents, *download* is good enough for now
- It should follow the same rules as to how the spaces are shared, e.g., whether it's from a personal space that belongs to the current user, or a shared space that the current user can access. 
- Right now files are stored on disk, and let's make it work against that setup for now; it will inevitably change in the future. 

See if you could make it work with the current architecture without adding any extra component first :)

**Comment by osala-eng** [comment]
> Great points. Let me clarify:
> 
> * No need to _view_ documents, _download_ is good enough for now
> * It should follow the same rules as to how the spaces are shared, e.g., whether it's from a personal space that belongs to the current user, or a shared space that the current user can access.
> * Right now files are stored on disk, and let's make it work against that setup for now; it will inevitably change in the future.
> 
> See if you could make it work with the current architecture without adding any extra component first :)

Okay :)
If only to make the files downloadable, this should be fine without any extra components.

**Comment by cwang** [comment]
BTW use rebase in branches for a cleaner history, see https://www.atlassian.com/git/tutorials/merging-vs-rebasing#:~:text=The%20Rebase%20Option&text=This%20moves%20the%20entire%20feature,commit%20in%20the%20original%20branch.

**Comment by janaka** [comment]
We should also try and follow conventional commits, I like it as a system to help me be consistent with commit messages as have sufficient concise context.

https://www.conventionalcommits.org/en/v1.0.0/

**Comment by cwang** [comment]
See https://github.com/docqai/docq/pull/41 and you may want to rebase to pull the changes from that PR once it's merged. 

What you're gonna get is something below, prior to message formatting:
Screenshot 2023-06-30 at 10 42 34

After reformatting,
- It should say "File: {Document Name}" linked to a downloadable file
- It should say "Page: {Page Number}" too

**Comment by cwang** [comment]
Also pls have a look at #41 and approve if you like @osala-eng

**Comment by osala-eng** [comment]
> Also pls have a look at #41 and approve if you like @osala-eng

Okay

**Comment by cwang** [comment]
@janaka wants to cut a release - would love to have this source display feature included - @osala-eng could we get it done?

**Comment by osala-eng** [comment]
After formatting and adding a downloadable link

sources

This still needs a bit more styling for the `blockquote` element from the `st-chat` component

**Comment by cwang** [comment]
> After formatting and adding a downloadable link
> 
> sources
> This still needs a bit more styling for the `blockquote` element from the `st-chat` component

Agreed, would be better if it's more compact

**Comment by osala-eng** [comment]
> After formatting and adding a downloadable link
> 
> sources
> This still needs a bit more styling for the `blockquote` element from the `st-chat` component

Should I combine all the sources refering to a similar document and have something like this instead:

> File: [Link150 user guide.pdf](https://web.net) <br>Pages: 21, 31, 48

**Comment by cwang** [comment]
> > After formatting and adding a downloadable link
> > sources
> > This still needs a bit more styling for the `blockquote` element from the `st-chat` component
> 
> Should I combine all the sources refering to a similar document and have something like this instead:
> 
> > File: [Link150 user guide.pdf](https://web.net) Pages: 21, 31, 48

I'd love it!!

File: source/docq/manage_documents.py [hunk 1]
```diff
@@ -6,6 +6,13 @@
 from .domain import SpaceKey
 from .support.llm import reindex
 from .support.store import get_upload_dir, get_upload_file
+from streamlit import runtime
+from mimetypes import guess_type
+from datetime import datetime
+import re
+from llama_index.utils import truncate_text
+import logging as log
+from llama_index.data_structs.node import NodeWithScore
```
Comment (cwang) [inline source/docq/manage_documents.py:15]: If you're using VC Code or another IDE, make sure to turn on Linting which should pick up the rules set up in `pyproject.toml` automatically (optionally set it up as triggered on save). Pointing it out here as organising imports is one of the things a Linter does really well.

File: source/docq/run_queries.py [hunk 2]
```diff
@@ -96,7 +97,7 @@ def query(input_: str, feature: FeatureKey, space: SpaceKey, spaces: list[SpaceK
             MESSAGE_TEMPLATE.format(message=response.response)
             if is_chat
             else MESSAGE_WITH_SOURCES_TEMPLATE.format(
-                message=response.response, source=response.get_formatted_sources()
+                message=response.response, source=format_document_sources(response.source_nodes, space)
```
Comment (cwang) [inline source/docq/run_queries.py:100]: Maybe refactor this a little. The idea is that in the source module, we should avoid dealing with UI/UX stuff such as formatting, which should be done in the web part (with streamlit)

File: web/utils/layout.py [hunk 3]
```diff
@@ -177,7 +177,7 @@ def chat_ui(feature: FeatureKey) -> None:
             if format_datetime(time) != day:
                 day = format_datetime(time)
                 st.markdown(f"#### {day}")
-            message(text, is_user, key=f"{key}_{feature.value()}")
+            message(text, is_user, key=f"{key}_{feature.value()}", allow_html=True)
```
Comment (cwang) [inline web/utils/layout.py:180]: Did they (st-chat) choose the same `allow_html` arg to enable HTML, or you missed this one :)?

**Review by cwang — COMMENTED** [review]
As I mentioned, formatting is best to be on the web side instead of living in the source module.

File: source/docq/manage_documents.py (hunk 1 above)
Comment (janaka) [reply source/docq/manage_documents.py:15]: which linter are we using?

Comment (cwang) [reply source/docq/manage_documents.py:15]: Ruff and its VC code extension https://marketplace.visualstudio.com/items?itemName=charliermarsh.ruff

File: web/utils/layout.py (hunk 3 above)
Comment (osala-eng) [reply web/utils/layout.py:180]: This is the same for st-chat as well.

**Comment by cwang** [comment]
Also @osala-eng let's switch to use https://docs.streamlit.io/library/api-reference/chat which is the latest UI elements from streamlit and a lot cleaner than st-chat.

**Comment by osala-eng** [comment]
> Also @osala-eng let's switch to use https://docs.streamlit.io/library/api-reference/chat which is the latest UI elements from streamlit and a lot cleaner than st-chat.

Okay, let me have a look at it.

File: source/docq/manage_documents.py [hunk 4]
```diff
@@ -38,4 +46,50 @@ def delete_all(space: SpaceKey) -> None:
...
+                else:
+                    source_groups[file_name] = [page_label]
+        except Exception as e:
+            log.exception("Error formatting source %s", e)
+            continue
+
+    for file_name, page_labels in source_groups.items():
+        download_url = _get_download_link(file_name, space)
+        _sources.append(f"> *File:* [{file_name}]({download_url})<br> *Pages:* {', '.join(page_labels)}")
+    return delimitter.join(_sources)
```
Comment (cwang) [inline source/docq/manage_documents.py:95]: I'm surprised that we have to parse text to get to these data points but I guess it's what it is :)

File: source/docq/run_queries.py (hunk 2 above)
Comment (cwang) [reply source/docq/run_queries.py:101]: Let's leave it here for now. Longer term this function should probably return data objects and then leave the UI layer to format but no need to shift focus now.

**Review by cwang — COMMENTED** [review]
Left one comment

**Comment by osala-eng** [comment]
> Also @osala-eng let's switch to use https://docs.streamlit.io/library/api-reference/chat which is the latest UI elements from streamlit and a lot cleaner than st-chat.

I've modified the chat-ui to use `st.chat_message` for the `AI` response.

chat-page

File: web/utils/layout.py [hunk 5]
```diff
@@ -157,7 +156,15 @@ def list_users_ui(username_match: str = None) -> None:
                         st.form_submit_button("Save", on_click=handle_update_user, args=(id_,))
 
 
+def _chat_message(message_: str, is_user: bool, key: str) -> None:
+    if is_user:
+        message(message_, key=key, is_user=True)
```
Comment (cwang) [inline web/utils/layout.py:161]: `streamlit_chat` is still used here, intentionally?

File: tests/docq/manage_documents_test.py [hunk 6]
```diff
@@ -0,0 +1,62 @@
...
+node_text = """
+page_label: 0
+file_name: test.txt
+Mock text
+"""
+mock_node = Node(node_text, "test", [1, 2, 3], "test")
+mock_node_with_score = NodeWithScore(mock_node, 0.5)
+
+
+@pytest.fixture(scope="module", autouse=True)
```
Comment (cwang) [inline tests/docq/manage_documents_test.py:23]: I used a class structure in https://github.com/docqai/docq/blob/main/tests/docq/support/store_test.py as an alternative. No need to change here.

File: tests/docq/manage_documents_test.py [hunk 7]
```diff
@@ -0,0 +1,62 @@
...
+    ("source_nodes", "space", "expected"),
+    [
+        ([], "personal_1234", ""),
+        ([mock_node_with_score], "personal_1234", '> *File:* [test.txt]()<br> *Pages:* 0'),
+        (["failing-node"], "personal_1234", "")
+    ]
+)
+def test_format_document_sources(source_nodes: list[NodeWithScore], space: SpaceKey, expected: str) -> None:
+    """Test that the document sources are formatted correctly."""
+    assert format_document_sources(source_nodes, space) == expected
```
Comment (cwang) [inline tests/docq/manage_documents_test.py:62]: Great

**Review by cwang — COMMENTED** [review]
A few comments

File: web/utils/layout.py (hunk 5 above)
Comment (osala-eng) [reply web/utils/layout.py:161]: Yes, this is intentional. `st.chat_message` aligns all the messages on the same side making it difficult to distinguish between user and assistant messages

Comment (cwang) [reply web/utils/layout.py:161]: Can we do something about it, e.g. supplying custom CSS to change the alignment? I feel that it defeats the purpose of switching to it if we end up using both. If we don't have any way of making it work then we should stick to `streamlit-chat` for now. 

BTW would be nice if this is raised with the latest commit - it's a significant discovery.

Comment (cwang) [reply web/utils/layout.py:161]: @janaka what do you think re NOT doing left/right alignment on robot/human messages? It was the default with `streamlit-chat` but as long as we use different avatars, the chatgpt style, non-alignment UI is okay?

Comment (janaka) [reply web/utils/layout.py:161]: I don't see all left as an issue as long as the avatar is clear between user ("you") and bot ("Docq" ?).

Comment (cwang) [reply web/utils/layout.py:161]: @osala-eng don't worry about the alignment but do keep differentiating between user and bot, i.e. by using different avatars

Comment (janaka) [reply web/utils/layout.py:161]: `st.chat_message` and `st.chat_input` are the the chat components though right ?

Comment (osala-eng) [reply web/utils/layout.py:161]: okay

Comment (janaka) [reply web/utils/layout.py:161]: @osala-eng 
These default avatars are fine as place holders to get going, we'll decide on something more custom afterwards.

image

If you've not seen this doc:
https://docs.streamlit.io/knowledge-base/tutorials/build-conversational-apps

**Comment by osala-eng** [comment]
The new layout

chat-layout

**Review by cwang — APPROVED** [review]

**Pull request merged**